File: common-api/internal/model/room.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type Room struct {
	ID          uuid.UUID `json:"id"`
	Name        string    `json:"name"`
	Code        string    `json:"code"`
	RoomType    string    `json:"room_type"`
	Capacity    int32     `json:"capacity"`
	Floor       *int32    `json:"floor,omitempty"`
	Building    *string   `json:"building,omitempty"`
	Description *string   `json:"description,omitempty"`
	Amenities   []string  `json:"amenities"`
	Status      string    `json:"status"`
	ImageURL    *string   `json:"image_url,omitempty"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

type CreateRoomRequest struct {
	Name        string   `json:"name" binding:"required"`
	Code        string   `json:"code" binding:"required"`
	RoomType    string   `json:"room_type" binding:"required,oneof=classroom meeting_room lab auditorium"`
	Capacity    int32    `json:"capacity" binding:"required,min=1"`
	Floor       *int32   `json:"floor,omitempty"`
	Building    *string  `json:"building,omitempty"`
	Description *string  `json:"description,omitempty"`
	Amenities   []string `json:"amenities,omitempty"`
	Status      *string  `json:"status,omitempty" binding:"omitempty,oneof=available maintenance closed"`
}

type UpdateRoomRequest struct {
	Name        *string  `json:"name,omitempty"`
	Code        *string  `json:"code,omitempty"`
	RoomType    *string  `json:"room_type,omitempty" binding:"omitempty,oneof=classroom meeting_room lab auditorium"`
	Capacity    *int32   `json:"capacity,omitempty" binding:"omitempty,min=1"`
	Floor       *int32   `json:"floor,omitempty"`
	Building    *string  `json:"building,omitempty"`
	Description *string  `json:"description,omitempty"`
	Amenities   []string `json:"amenities,omitempty"`
	Status      *string  `json:"status,omitempty" binding:"omitempty,oneof=available maintenance closed"`
}

File: web/lib/schemas/room.ts
import { z } from 'zod'

export const roomSchema = z.object({
  name: z.string().min(1, 'Name is required'),
  code: z.string().min(1, 'Code is required'),
  room_type: z.enum(['classroom', 'meeting_room', 'lab', 'auditorium', 'other']),
  capacity: z.coerce.number().int().min(1, 'Capacity must be at least 1'),
  floor: z.coerce.number().int().optional(),
  building: z.string().optional(),
  description: z.string().optional(),
  amenities: z.array(z.string()).optional(),
  status: z.enum(['available', 'maintenance', 'closed']).optional(),
})

export type RoomInput = z.infer<typeof roomSchema>
